clippy_lints/src/single_field_patterns.rs

```diff
@@ -0,0 +1,264 @@
+    ///     None => {},
+    ///     _ => {},
+    /// }
+    /// # }
+    /// ```
+    #[clippy::version = "1.59.0"]
+    pub SINGLE_FIELD_PATTERNS,
+    style,
+    "single-field patterns"
+}
+declare_lint_pass!(SingleFieldPatterns => [SINGLE_FIELD_PATTERNS]);
+
+/// This represents 0 or 1 fields being used. Where more may be used, I use Option<SingleField>
+///   where None represents the absence of a lint
+#[derive(Debug, Clone, Copy)]
+enum SingleField {
+    Id { id: Ident, pattern: Span },
+    Unused, // The name "SingleField" is a lie but idk what's better. "AtMostOneField"?
+}
+
+impl PartialEq for SingleField {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (SingleField::Id { id: id1, .. }, SingleField::Id { id: id2, .. }) => id1 == id2,
+            (SingleField::Unused, SingleField::Unused) => true,
+            _ => false,
+        }
+    }
+}
+
+fn get_sf<'a>(mut iter: impl Iterator<Item = (Ident, &'a Pat<'a>)>) -> Option<SingleField> {
+    let one = iter.by_ref().find(|(_, pat)| !matches!(pat.kind, PatKind::Wild));
+    match one {
+        Some((id, pat)) => {
+            if pat.span.from_expansion() {
+                return None;
+            }
+            if iter.all(|(_, other)| matches!(other.kind, PatKind::Wild)) {
+                Some(SingleField::Id { id, pattern: pat.span })
+            } else {
+                None
+            }
+        },
+        None => Some(SingleField::Unused),
+    }
+}
+
+fn struct_sf(pat: &PatKind<'_>) -> Option<SingleField> {
+    match *pat {
+        PatKind::Struct(_, pats, _) => get_sf(pats.iter().map(|field| (field.ident, field.pat))),
+        _ => None,
+    }
+}
+
+/// This handles recursive patterns and flattens them out lazily
+/// e.g. 1 | (2 | 9) | 3..5
+struct FlatPatterns<'hir, I>
+where
+    I: Iterator<Item = &'hir Pat<'hir>>,
+{
+    patterns: I,
+    stack: Vec<&'hir Pat<'hir>>,
+}
+
+impl<I: Iterator<Item = &'hir Pat<'hir>>> FlatPatterns<'hir, I> {
+    fn new(patterns: I) -> Self {
+        Self {
+            patterns,
+            stack: Vec::new(),
+        }
+    }
+}
+
+impl<I: Iterator<Item = &'hir Pat<'hir>>> Iterator for FlatPatterns<'hir, I> {
+    type Item = &'hir Pat<'hir>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.stack.is_empty() {
+            if let Some(pat) = self.patterns.next() {
+                self.stack.push(pat);
+            } else {
+                return None;
+            }
+        }
+        while let Some(pat) = self.stack.pop() {
+            match pat.kind {
+                PatKind::Or(pats) => self.stack.extend(pats),
+                _ => return Some(pat),
+            }
+        }
+        unreachable!("Or always has 2 patterns, so one of the prior returns must return");
+    }
+}
+
+fn find_sf_lint<'hir>(
+    cx: &LateContext<'_>,
+    patterns: impl Iterator<Item = &'hir Pat<'hir>>,
+    leaf_sf: &impl Fn(&PatKind<'hir>) -> Option<SingleField>,
```

Comment:
Why is this passed? There's only one place this function is called. Probably a relic of a previous version?